vaadin-renderer-flow-parent/vaadin-renderer-flow-integration-tests/test/flow-component-renderer.test.ts

```diff
@@ -67,4 +89,94 @@ describe('flow-component-renderer', () => {
+    const component = fixtureSync<TestComponent>(`<test-component></test-component>`);
+
+    // Create an element to render
+    elements[0] = document.createElement('button');
+    component.nodeId = 0;
+    await nextFrame();
+
+    // Remove the element (simulate removing from the Flow's registry)
+    delete elements[0];
+
+    // Render once again with the old node id (this can happen if Grid's items haven't
+    // yet been updated not to include the removed node id and re-render is invoked)
+    component.requestUpdate();
+    await nextFrame();
+
+    getByNodeId.resetHistory();
+    // Finally, render with undefined node id (the Grid's items have been updated)
+    component.nodeId = undefined;
+
+    await nextFrame();
+    expect(getByNodeId).to.not.have.been.called;
+    expect(component.firstElementChild).to.equal(null);
+  });
+
+  it('should not try to re-render a replaced node', async () => {
+    const { getByNodeId } = window.Vaadin.Flow.clients.ROOT;
+    const component = fixtureSync<TestComponent>(`<test-component></test-component>`);
+
+    // Create an element to render
+    elements[0] = document.createElement('button');
+    component.nodeId = 0;
+    await nextFrame();
+
+    // Remove the element
+    delete elements[0];
+    // Add new element
+    elements[1] = document.createElement('button');
+    // Render once again with the old node id
+    component.requestUpdate();
+    await nextFrame();
+
+    getByNodeId.resetHistory();
+    // Finally, render with the new node id
+    component.nodeId = 1;
+    await nextFrame();
+
+    // getByNodeId should only have been called for node id 1 only
+    getByNodeId.getCalls().forEach((call) => expect(call.args[0]).to.equal(1));
+    expect(component.firstElementChild).to.equal(elements[1]);
+  });
+
+  it('should not replace new synchronous node with the old lazily added node', async () => {
+    const component = fixtureSync<TestComponent>(`<test-component></test-component>`);
+
+    // Try to render a lazily added node (not added to the Flow's registry yet)
+    component.nodeId = 0;
+    await nextFrame();
+
+    // Add and render a new node synchronously
+    elements[1] = document.createElement('button');
+    component.nodeId = 1;
+    await nextFrame();
+
+    // Add the lazy node to the Flow's registry
+    elements[0] = document.createElement('button');
+    await nextFrame();
+
+    // The new node should still be rendered
+    expect(component.firstElementChild).to.equal(elements[1]);
+    expect(component.contains(elements[0])).to.be.false;
+  });
+
+  it('should not try to re-render an old node after disconnect', async () => {
+    const { getByNodeId } = window.Vaadin.Flow.clients.ROOT;
+    const component = fixtureSync<TestComponent>(`<test-component></test-component>`);
+
+    // Try to render a lazily added node
+    component.nodeId = 0;
+    await nextFrame();
+
+    // Disconnect
+    component.remove();
+    getByNodeId.resetHistory();
```

Comment:
Should it be before `component.remove()`?

Reply:
The order doesn't seem to make a difference. Maybe it could be changed for clarity.

Reply:
Yeah, I'd suggest changing it to avoid confusion.

Reply:
Done